File: src/domain/usecases/users/ChangeOwnPasswordUseCase.ts
import type { AuthRepository } from '@/domain/repositories/AuthRepository'
import type { UserRepository } from '@/domain/repositories/UserRepository'
import type { User } from '@/domain/entities/User'
import {
  UnauthorizedError,
  ValidationError,
} from '@/domain/errors/DomainError'
import {
  DEFAULT_TEMPORARY_PASSWORD,
  isSecurePassword,
  securePasswordRequirementsMessage,
} from '@/domain/value-objects/PasswordPolicy'

export class ChangeOwnPasswordUseCase {
  private readonly authRepository: AuthRepository
  private readonly userRepository: UserRepository

  constructor(authRepository: AuthRepository, userRepository: UserRepository) {
    this.authRepository = authRepository
    this.userRepository = userRepository
  }

  async execute(
    actor: User,
    newPassword: string,
    confirmPassword: string,
  ): Promise<User> {
    if (!actor.active) {
      throw new UnauthorizedError('Cuenta inactiva')
    }

    if (!actor.mustChangePassword) {
      throw new ValidationError('Tu contraseña ya fue actualizada')
    }

    if (!newPassword || !confirmPassword) {
      throw new ValidationError('Completa ambos campos de contraseña')
    }

    if (newPassword !== confirmPassword) {
      throw new ValidationError('Las contraseñas no coinciden')
    }

    if (!isSecurePassword(newPassword)) {
      throw new ValidationError(securePasswordRequirementsMessage())
    }

    if (newPassword === DEFAULT_TEMPORARY_PASSWORD) {
      throw new ValidationError('Debes elegir una contraseña distinta a la temporal')
    }

    await this.authRepository.updatePassword(newPassword)

    return this.userRepository.update(actor.id, {
      mustChangePassword: false,
    })
  }
}

File: consorcio/lib/domain/usecases/change_own_password_use_case.dart
import '../entities/app_user.dart';
import '../errors/domain_exception.dart';
import '../repositories/auth_repository.dart';
import '../repositories/user_repository.dart';
import '../value_objects/password_policy.dart';

class ChangeOwnPasswordUseCase {
  ChangeOwnPasswordUseCase(this._authRepository, this._userRepository);

  final AuthRepository _authRepository;
  final UserRepository _userRepository;

  Future<AppUser> execute({
    required AppUser actor,
    required String newPassword,
    required String confirmPassword,
  }) async {
    if (!actor.active) {
      throw DomainException('Cuenta inactiva');
    }

    if (!actor.mustChangePassword) {
      throw DomainException('Tu contraseña ya fue actualizada');
    }

    if (newPassword.isEmpty || confirmPassword.isEmpty) {
      throw DomainException('Completa ambos campos de contraseña');
    }

    if (newPassword != confirmPassword) {
      throw DomainException('Las contraseñas no coinciden');
    }

    if (!isSecurePassword(newPassword)) {
      throw DomainException(securePasswordRequirementsMessage());
    }

    if (newPassword == defaultTemporaryPassword) {
      throw DomainException(
        'Debes elegir una contraseña distinta a la temporal',
      );
    }

    await _authRepository.updatePassword(newPassword);

    return _userRepository.updateMustChangePassword(
      userId: actor.id,
      mustChangePassword: false,
    );
  }
}
